website/docs/d/google_impersonated_credential.html.markdown

```diff
@@ -0,0 +1,84 @@
+---
+layout: "google"
+page_title: "Google: google_impersonated_credential"
+sidebar_current: "docs-google-impersonated-credential"
+description: |-
+  Produces access_token for impersonated service accounts
+---
+
+# google\_impersonated\_credential
+
+This data source provides a google `oauth2` `access_token` for a different service account than the one initially running the script.  You can 
+then use this new token to access resources the original caller would not have permissions on otherwise.
+
+For more information see
+[the official documentation](https://cloud.google.com/iam/docs/creating-short-lived-service-account-credentials) as well as [iamcredentials.generateAccessToken()](https://cloud.google.com/iam/credentials/reference/rest/v1/projects.serviceAccounts/generateAccessToken)
+
+## Example Usage
+
+To allow `service_A` to impersonate `service_B`, grant the [Service Account Token Creator](https://cloud.google.com/iam/docs/service-accounts#the_service_account_token_creator_role) on B to A. 
+
+In the IAM policy below, `service_A` is given the Token Creator role impersonate `service_B`
+
+```hcl
+$ cat service_policy.json 
+{ 
+"bindings": [
+    {
+      "members": [
+        "service_A@projectA.iam.gserviceaccount.com "
+      ],
+      "role": "roles/iam.serviceAccountTokenCreator",    
+    }
+  ]
+}
+
+$ gcloud iam service-accounts set-iam-policy  service_B@projectB.iam.gserviceaccount.com  service_policy.json
+```
+
+Once the IAM permissions are set, you can apply the new token to a provider bootstrapped with it.  Any resources that references the new provider will run as the new identity.
+
+In the example below, `google_project` will run as `service_B`.
+
+```hcl
+provider "google" {}
+
+data "google_client_config" "default" {
+  provider = "google"
+}
+
+data "google_impersonated_credential" "default" {
+ provider = "google"
+ target_service_account = "service_B@projectB.iam.gserviceaccount.com"
+ scopes = ["devstorage.read_only", "cloud-platform"]
+ lifetime = "300s"
+}
+
+provider "google" {
+   alias  = "impersonated"
+   access_token = "${data.google_impersonated_credential.default.access_token}"
+}
+
+data "google_project" "project" {
+  provider = "google.impersonated"
+  project_id = "target-project"
+}
+```
+
+> *Note*: the generated token is non-refreshable and can have a maximum `lifetime` of `3600` seconds.
```

Comment:
The token is also non-revocable, which is my biggest objection to this process overall. If you create an oauth2 access token that's valid for 3600 seconds, it lives for one hour and cannot be revoked. You'd need to revoke the parent service account key (or parent service account), which could impact multiple services.

Reply:
all service accounts tokens are non-revokeable (the token itself, that is).  The svc account that  terraform bootstraps with or the one time token provided by [GOOGLE_OAUTH_ACCESS_TOKEN](https://www.terraform.io/docs/providers/google/provider_reference.html#configuration-reference) if its derived from a svc_account isn't revokeable anyway for its duration.

[user tokens](https://developers.google.com/identity/protocols/OAuth2WebServer#tokenrevoke) are revokeable, however

the concern of non-refreshable svc tokens is something different though (and is preexisting already).   if the objection is that that we infact do want to have a `refreshable` token, that is something else and debateable (we'd need to pass in an actual google oauth2 renewable credential in and i'd have to bake in logic similar to this in this cancelled PR or [oauth2/google](https://go-review.googlesource.com/c/oauth2/+/143957/13/google/impersonate.go)

@sethvargo  would you comment on the issue https://github.com/terraform-providers/terraform-provider-google/issues/3215  i opened for this?

Reply:
marking this reolved; LMK if there are follwup/reopen